// @runware/sdk-js/dist/esm/tests/test-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testExamples = exports.mockUploadFile = exports.mockNewImageUID = exports.mockFileToBase64 = exports.mockTextImageUpload = exports.mockTaskUUID = void 0;
const Runware_1 = require("../Runware");
const utils_1 = require("../Runware/utils");
const promptText = "A beautiful runware";
exports.mockTaskUUID = "UNIQUE_UID";
exports.mockTextImageUpload = "IMAGE_UPLOAD";
exports.mockFileToBase64 = "FILE_TO_BASE_64";
exports.mockNewImageUID = "NEW_IMAGE_UID";
exports.mockUploadFile = new utils_1.MockFile().create("pic.jpg", 1024 * 1024 * 2, "image/jpeg");
exports.testExamples = {
    imageReq: {
        numberResults: 8,
        positivePrompt: promptText,
        model: 13,
        steps: 30,
        width: 512,
        height: 512,
    },
    imageRes: {
        model: 13,
        numberResults: 8,
        positivePrompt: promptText,
        steps: 30,
        taskType: Runware_1.ETaskType.IMAGE_INFERENCE,
        taskUUID: exports.mockTaskUUID,
        width: 512,
        height: 512,
    },
    imageUploadRes: {
        imageUUID: exports.mockNewImageUID,
        imageURL: "data:image/png;base64,iVBORw0KGgoAAAA...",
        taskUUID: "50836053-a0ee-4cf5-b9d6-ae7c5d140ada",
    },
    controlNet: {
        endStep: 20,
        startStep: 0,
        guideImage: exports.mockTextImageUpload,
        preprocessor: "canny",
        weight: 1,
        controlMode: Runware_1.EControlMode.CONTROL_NET,
    },
};
