// @runware/sdk-js/dist/esm/Runware/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EOpenPosePreProcessor = exports.EPreProcessor = exports.EPreProcessorGroup = exports.EControlMode = exports.ETaskType = exports.SdkType = exports.Environment = void 0;
var Environment;
(function (Environment) {
    Environment["PRODUCTION"] = "PRODUCTION";
    Environment["DEVELOPMENT"] = "DEVELOPMENT";
    Environment["TEST"] = "TEST";
})(Environment || (exports.Environment = Environment = {}));
var SdkType;
(function (SdkType) {
    SdkType["CLIENT"] = "CLIENT";
    SdkType["SERVER"] = "SERVER";
})(SdkType || (exports.SdkType = SdkType = {}));
var ETaskType;
(function (ETaskType) {
    ETaskType["IMAGE_INFERENCE"] = "imageInference";
    ETaskType["IMAGE_UPLOAD"] = "imageUpload";
    ETaskType["IMAGE_UPSCALE"] = "imageUpscale";
    ETaskType["IMAGE_BACKGROUND_REMOVAL"] = "imageBackgroundRemoval";
    ETaskType["IMAGE_CAPTION"] = "imageCaption";
    ETaskType["IMAGE_CONTROL_NET_PRE_PROCESS"] = "imageControlNetPreProcess";
    ETaskType["PROMPT_ENHANCE"] = "promptEnhance";
    ETaskType["AUTHENTICATION"] = "authentication";
})(ETaskType || (exports.ETaskType = ETaskType = {}));
var EControlMode;
(function (EControlMode) {
    EControlMode["BALANCED"] = "balanced";
    EControlMode["PROMPT"] = "prompt";
    EControlMode["CONTROL_NET"] = "controlnet";
})(EControlMode || (exports.EControlMode = EControlMode = {}));
var EPreProcessorGroup;
(function (EPreProcessorGroup) {
    EPreProcessorGroup["canny"] = "canny";
    EPreProcessorGroup["depth"] = "depth";
    EPreProcessorGroup["mlsd"] = "mlsd";
    EPreProcessorGroup["normalbae"] = "normalbae";
    EPreProcessorGroup["openpose"] = "openpose";
    EPreProcessorGroup["tile"] = "tile";
    EPreProcessorGroup["seg"] = "seg";
    EPreProcessorGroup["lineart"] = "lineart";
    EPreProcessorGroup["lineart_anime"] = "lineart_anime";
    EPreProcessorGroup["shuffle"] = "shuffle";
    EPreProcessorGroup["scribble"] = "scribble";
    EPreProcessorGroup["softedge"] = "softedge";
})(EPreProcessorGroup || (exports.EPreProcessorGroup = EPreProcessorGroup = {}));
var EPreProcessor;
(function (EPreProcessor) {
    EPreProcessor["canny"] = "canny";
    EPreProcessor["depth_leres"] = "depth_leres";
    EPreProcessor["depth_midas"] = "depth_midas";
    EPreProcessor["depth_zoe"] = "depth_zoe";
    EPreProcessor["inpaint_global_harmonious"] = "inpaint_global_harmonious";
    EPreProcessor["lineart_anime"] = "lineart_anime";
    EPreProcessor["lineart_coarse"] = "lineart_coarse";
    EPreProcessor["lineart_realistic"] = "lineart_realistic";
    EPreProcessor["lineart_standard"] = "lineart_standard";
    EPreProcessor["mlsd"] = "mlsd";
    EPreProcessor["normal_bae"] = "normal_bae";
    EPreProcessor["scribble_hed"] = "scribble_hed";
    EPreProcessor["scribble_pidinet"] = "scribble_pidinet";
    EPreProcessor["seg_ofade20k"] = "seg_ofade20k";
    EPreProcessor["seg_ofcoco"] = "seg_ofcoco";
    EPreProcessor["seg_ufade20k"] = "seg_ufade20k";
    EPreProcessor["shuffle"] = "shuffle";
    EPreProcessor["softedge_hed"] = "softedge_hed";
    EPreProcessor["softedge_hedsafe"] = "softedge_hedsafe";
    EPreProcessor["softedge_pidinet"] = "softedge_pidinet";
    EPreProcessor["softedge_pidisafe"] = "softedge_pidisafe";
    EPreProcessor["tile_gaussian"] = "tile_gaussian";
    EPreProcessor["openpose"] = "openpose";
    EPreProcessor["openpose_face"] = "openpose_face";
    EPreProcessor["openpose_faceonly"] = "openpose_faceonly";
    EPreProcessor["openpose_full"] = "openpose_full";
    EPreProcessor["openpose_hand"] = "openpose_hand";
})(EPreProcessor || (exports.EPreProcessor = EPreProcessor = {}));
var EOpenPosePreProcessor;
(function (EOpenPosePreProcessor) {
    EOpenPosePreProcessor["openpose"] = "openpose";
    EOpenPosePreProcessor["openpose_face"] = "openpose_face";
    EOpenPosePreProcessor["openpose_faceonly"] = "openpose_faceonly";
    EOpenPosePreProcessor["openpose_full"] = "openpose_full";
    EOpenPosePreProcessor["openpose_hand"] = "openpose_hand";
})(EOpenPosePreProcessor || (exports.EOpenPosePreProcessor = EOpenPosePreProcessor = {}));
